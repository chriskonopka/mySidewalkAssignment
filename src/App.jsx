import React, { Fragment, useState, useEffect } from 'react';
import Map, {Source, Layer} from 'react-map-gl';
import Highcharts from 'highcharts'
import HighchartsReact from 'highcharts-react-official'
import accessibility from 'highcharts/modules/accessibility';

import 'mapbox-gl/dist/mapbox-gl.css';

accessibility(Highcharts);

const MAPBOX_TOKEN = 'pk.eyJ1IjoiY2prb25vcGthIiwiYSI6ImNscWZzcWd3NjEyMjIyaXFmZ3lzZDB6cGMifQ.GvJTjQf3TiTrvsUTdURoiw'; // Set your mapbox token here       
 
export function createNameLabel(shid) {
  // shid format puts the name at the end preded by a :
  // the type, neighborhood or tract, is immediately before that
  const parts = shid.split('/');
  const typeAndId = parts[parts.length-1].split(':');
  const type = typeAndId[0];
  const id = typeAndId[1];

  switch (type) {
    case 'tract':
      return `Tract: ${id}`;
    case 'neighborhood':
      return `Neighborhood: ${id.replace(/_/g, ' ')}`;
    default:
      return '';
  }
};

export function createChartOptions(selectedFeatures) {
  // used chatGPT to help format dataSeries input for highchart
  const categories = ['drive_alone', 'drive_carpool', 'public_transit', 'walk'];
  const dataSeries = selectedFeatures.map((feature) => {
    const shid = feature.properties.shid;
    const commuteData = categories.map((category) => feature.properties[`pop-commute-${category}`]);
    return [shid, ...commuteData];
  });

  return {
    chart: {
      type: 'bar',
      height: '300px',
      accessibility: {
        description: 'Bar chart displaying commuter populations in selected regions, broken down by type of commute',
      },
    },
    title: {
      text: 'Commuter Populations in Selected Regions',
    },
    xAxis: {
      categories: categories,
      title: {
        text: 'Commute type',
      },
    },
    yAxis: {
      title: {
        text: 'Number of Commuters',
      },
    },
    series: dataSeries.map((data) => ({
      // shid is the first entry in data, use it for name and remove it for commuter data
    name: createNameLabel(data[0]),
      data: data.slice(1),
      // Keep color consistancy with map colors
      color: data[0].includes('tract') > 0 ? '#800000' : '#808080',
    })),
  };
}

function App() {
  const [selectedFeatures, setSelectedFeatures] = useState([]);
  const [kcNeighborhoods, setKcNeighborhoods] = useState(null);
  const [kcTracts, setKcTracts] = useState(null);

  const chartOptions = createChartOptions(selectedFeatures)

  // empty dependancy array so data isn't constantly re-fetched
  useEffect(() => {
    const fetchGeoJSON = async (url, setFunction) => {
      try {
        const response = await fetch(url);
        const data = await response.json();
        setFunction(data);
      } catch (error) {
        console.error('Error fetching GeoJSON data from Repo:', error);
      }
    };
    fetchGeoJSON('https://raw.githubusercontent.com/mysidewalk/interview/master/frontend-engineer/kc-neighborhoods.json', setKcNeighborhoods);
    fetchGeoJSON('https://raw.githubusercontent.com/mysidewalk/interview/master/frontend-engineer/kc-tracts.json', setKcTracts);
  }, []);

  // Use serialIds for node tests so snapshots match exactly
  let testFlag = false;
  if (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'test') {
    testFlag = true;
  }
  Highcharts.useSerialIds(testFlag);

  const onMapClick = (event) => {
    if(event.features) {
      setSelectedFeatures(event.features)
    }
  }

  return (
    <Fragment>
      <HighchartsReact highcharts={Highcharts} options={chartOptions} />
    <Map
      // Inital react map code from react-map-gl getting started guide
      // https://visgl.github.io/react-map-gl/docs/get-started
      // longitude and latitude of Kansas City
      initialViewState={{
        latitude: 39.09,
        longitude: -94.57,
        zoom: 10
      }}
      id='KCMap'
      interactiveLayerIds={['neighborhoods','tracts'] }
      style={{width: '100%', height: 500}}
      mapStyle="mapbox://styles/mapbox/streets-v9"
      mapboxAccessToken={MAPBOX_TOKEN}
      onClick={onMapClick}
    >
      <Source type="geojson" data={kcNeighborhoods}>
        <Layer
          id="neighborhoods"
          type="fill"
          paint={{
            'fill-color': [
            'case',
            // I used chatGPT to structure this case syntax correctly
            // Code is used to change shading/opacity of region when selected
            ['in', ['get', 'id'], ['literal', selectedFeatures.map(f => f.properties.id)]],
            '#000000', // black when selected
            '#808080' // gray when not selected
          ],
            'fill-opacity': [
              'case',
              ['in', ['get', 'id'], ['literal', selectedFeatures.map(f => f.properties.id)]],
              .7,
              .5
            ],
          }}
        />
      </Source>
      <Source type="geojson" data={kcTracts}>
        <Layer
          id="tracts"
          type="fill"
          paint={{
            'fill-color': [
              'case',
              ['in', ['get', 'id'], ['literal', selectedFeatures.map(f => f.properties.id)]],
              '#FF0000', // red when selected
              '#800000' // maroon when unselected
            ],
            'fill-opacity': [
              'case',
              ['in', ['get', 'id'], ['literal', selectedFeatures.map(f => f.properties.id)]],
              .7,
              .5
            ],
          }}
        />
      </Source>
    </Map>
    </Fragment>
  );
}

export default App;
