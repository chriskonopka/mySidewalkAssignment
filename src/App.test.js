import { render, screen } from '@testing-library/react';
import App, {createNameLabel, createChartOptions} from './App';

// react-map-gl only works in a browser, needs to be mocked for the tests to run
jest.mock('react-map-gl', () => {
  const Map = jest.fn(({ children }) => <div>{children}</div>);
  const Source = jest.fn(({ children }) => <div>{children}</div>);
  const Layer = jest.fn(({ children }) => <div>{children}</div>);

  return {
    __esModule: true,
    default: Map,
    Source,
    Layer,
  };
});

jest.mock('highcharts/modules/accessibility', () => {
  return jest.fn();
});

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useEffect: jest.fn(),
}));

describe('Render', () => {
  it('renders the Chart Title', () => {
    render(<App />);
    const elements = screen.queryAllByText('Commuter Populations in Selected Regions')
    expect(elements[0]).toBeInTheDocument();
  
  });
  it('Matches DOM Snapshot', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
 });

describe('createNameLabel', () => {
  it('returns the correct label for a neighborhood shid', () => {
    const shid = 'country:us/state:mo/place:kansas_city/neighborhood:blue_vue_hills';
    const result = createNameLabel(shid);
    expect(result).toBe('Neighborhood: blue vue hills');
  });

  it('returns the correct label for a tract shid', () => {
    const shid = 'country:us/tract:29095012904';
    const result = createNameLabel(shid);
    expect(result).toBe('Tract: 29095012904');
  });

  it('returns an empty string for an unknown shid type', () => {
    const shid = 'country:us/state:mo/something:unknown';
    const result = createNameLabel(shid);
    expect(result).toBe('');
  });
});

describe('createChartOptions', () => {
  it('returns the correct chart options with two selected features', () => {
    const selectedFeatures = [
      {
          "type": "Feature",
          "state": {},
          "properties": {
              "id": 599128,
              "shid": "country:us/tract:29095015400",
              "area": 1766792,
              "pop-commute-drive_alone": 353,
              "pop-commute-drive_carpool": 43,
              "pop-commute-public_transit": 185,
              "pop-commute-walk": 48
          },
          "source": "jsx-source-3"
      },
      {
          "type": "Feature",
          "state": {},
          "properties": {
              "id": 935683,
              "shid": "country:us/state:mo/place:kansas_city/neighborhood:independence_plaza",
              "area": 1112740,
              "pop-commute-drive_alone": 483.2832189593,
              "pop-commute-drive_carpool": 117.361623342168,
              "pop-commute-public_transit": 109.537687136103,
              "pop-commute-walk": 45.3862513130519
          },
          "source": "jsx-source-1"
      }
    ]
    const expected = {
      "chart": {
        "accessibility": {
          "description": "Bar chart displaying commuter populations in selected regions, broken down by type of commute"
        }, 
        "height": "300px",
        "type": "bar"
      }, 
      "series": [
          {
            "color": "#800000",
            "data":[
              353,
              43,
              185,
              48,
            ],
            "name": "Tract: 29095015400",
          },
          {
            "color": "#808080",
            "data": [
              483.2832189593,
              117.361623342168,
              109.537687136103,
              45.3862513130519,
            ],
            "name": "Neighborhood: independence plaza",
          },
        ],
      "title": {
        "text": "Commuter Populations in Selected Regions"
      },
      "xAxis": {
        "categories": ["drive_alone", "drive_carpool", "public_transit", "walk"], 
        "title": {
          "text": "Commute type"
        }
      },
      "yAxis": {
        "title": {
          "text": "Number of Commuters"
        }
      }
    }
    const result = createChartOptions(selectedFeatures)
    expect(result).toEqual(expected);
  });

  it('returns the correct chart options with one selected feature', () => {
    const selectedFeatures = [
      {
          "type": "Feature",
          "state": {},
          "properties": {
              "id": 599128,
              "shid": "country:us/tract:29095015400",
              "area": 1766792,
              "pop-commute-drive_alone": 353,
              "pop-commute-drive_carpool": 43,
              "pop-commute-public_transit": 185,
              "pop-commute-walk": 48
          },
          "source": "jsx-source-3"
      },
      {
          "type": "Feature",
          "state": {},
          "properties": {
              "id": 935683,
              "shid": "country:us/state:mo/place:kansas_city/neighborhood:independence_plaza",
              "area": 1112740,
              "pop-commute-drive_alone": 483.2832189593,
              "pop-commute-drive_carpool": 117.361623342168,
              "pop-commute-public_transit": 109.537687136103,
              "pop-commute-walk": 45.3862513130519
          },
          "source": "jsx-source-1"
      }
    ]
    const expected = {
      "chart": {
        "accessibility": {
          "description": "Bar chart displaying commuter populations in selected regions, broken down by type of commute"
        }, 
        "height": "300px",
        "type": "bar"
      }, 
      "series": [
        {
          "color": "#800000",
          "data": [
            353,
            43,
            185,
            48,
          ],
          "name": "Tract: 29095015400",
        },
        {
          "color": "#808080",
          "data": [
            483.2832189593,
            117.361623342168,
            109.537687136103,
            45.3862513130519,
          ],
          "name": "Neighborhood: independence plaza",
        },
      ],
      "title": {
        "text": "Commuter Populations in Selected Regions"
      },
      "xAxis": {
        "categories": ["drive_alone", "drive_carpool", "public_transit", "walk"], 
        "title": {
          "text": "Commute type"
        }
      },
      "yAxis": {
        "title": {
          "text": "Number of Commuters"
        }
      }
    }
    const result = createChartOptions(selectedFeatures)
    expect(result).toEqual(expected);
  });

  it('returns the correct chart options with no selected features', () => {
    const selectedFeatures = []
    const expected = {
      "chart": {
        "accessibility": {
          "description": "Bar chart displaying commuter populations in selected regions, broken down by type of commute"
        }, 
        "height": "300px",
        "type": "bar"
      }, 
      "series": [], 
      "title": {
        "text": "Commuter Populations in Selected Regions"
      },
      "xAxis": {
        "categories": ["drive_alone", "drive_carpool", "public_transit", "walk"], 
        "title": {
          "text": "Commute type"
        }
      },
      "yAxis": {
        "title": {
          "text": "Number of Commuters"
        }
      }
    }
    const result = createChartOptions(selectedFeatures)
    expect(result).toEqual(expected);
  });
});