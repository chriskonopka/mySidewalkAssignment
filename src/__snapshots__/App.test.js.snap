// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Matches DOM Snapshot 1`] = `
<DocumentFragment>
  <div
    data-highcharts-chart="1"
    style="overflow: hidden;"
  >
    <div
      class="highcharts-container "
      dir="ltr"
      id="highcharts-2"
      style="position: relative; overflow: hidden; width: 600px; height: 300px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none;"
    >
      <svg
        aria-label="Commuter Populations in Selected Regions"
        class="highcharts-root"
        height="300"
        role="img"
        style="font-family: Helvetica, Arial, sans-serif; font-size: 1rem;"
        version="1.1"
        viewBox="0 0 600 300"
        width="600"
        xmlns="http://www.w3.org/2000/svg"
      >
        <desc>
          Created with Highcharts 11.2.0
        </desc>
        <defs>
          <filter
            id="highcharts-drop-shadow-1"
          >
            <fedropshadow
              dx="1"
              dy="1"
              flood-color="#000000"
              flood-opacity="0.75"
              stdDeviation="2.5"
            />
          </filter>
          <clippath
            id="highcharts-3-"
          >
            <rect
              fill="none"
              height="580"
              width="275"
              x="0"
              y="0"
            />
          </clippath>
        </defs>
        <rect
          class="highcharts-background"
          fill="#ffffff"
          filter="none"
          height="300"
          rx="0"
          ry="0"
          width="600"
          x="0"
          y="0"
        />
        <rect
          class="highcharts-plot-background"
          fill="none"
          filter="none"
          height="275"
          width="580"
          x="10"
          y="10"
        />
        <rect
          class="highcharts-plot-border"
          data-z-index="1"
          fill="none"
          height="275"
          stroke="#cccccc"
          stroke-width="0"
          width="580"
          x="10"
          y="10"
        />
        <g
          class="highcharts-series-group"
          data-z-index="3"
          filter="none"
        />
        <text
          class="highcharts-title"
          data-z-index="4"
          style="font-size: 1.2em; color: rgb(51, 51, 51); font-weight: bold; fill: #333333;"
          text-anchor="middle"
          x="300"
          y="10"
        >
          Commuter Populations in Selected Regions
        </text>
        <text
          class="highcharts-subtitle"
          data-z-index="4"
          style="color: rgb(102, 102, 102); font-size: 0.8em; fill: #666666;"
          text-anchor="middle"
          x="300"
          y="10"
        />
        <text
          class="highcharts-caption"
          data-z-index="4"
          style="color: rgb(102, 102, 102); font-size: 0.8em; fill: #666666;"
          text-anchor="start"
          x="10"
          y="287"
        />
        <g
          class="highcharts-legend highcharts-no-tooltip"
          data-z-index="7"
          visibility="hidden"
        >
          <rect
            class="highcharts-legend-box"
            fill="none"
            filter="none"
            height="8"
            rx="0"
            ry="0"
            stroke="#999999"
            stroke-width="0"
            width="8"
            x="0"
            y="0"
          />
          <g
            data-z-index="1"
          >
            <g />
          </g>
        </g>
        <text
          class="highcharts-credits"
          data-z-index="8"
          style="cursor: pointer; color: rgb(153, 153, 153); font-size: 0.6em; fill: #999999;"
          text-anchor="end"
          x="590"
          y="295"
        >
          Highcharts.com
        </text>
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
